core/src/banking_stage/consumer.rs

```diff
@@ -472,14 +472,23 @@ impl Consumer {
             ..
         } = execute_and_commit_transactions_output;
 
+        // Costs of all transactions are added before processing.
+        // To ensure accurate tracking of compute units, transactions that ultimately
+        // were not included in the block should have their cost removed.
+        QosService::remove_costs(
+            transaction_qos_cost_results.iter(),
+            commit_transactions_result.as_ref().ok(),
+            bank,
+        );
+
         // once feature `apply_cost_tracker_during_replay` is activated, leader shall no longer
         // adjust block with executed cost (a behavior more inline with bankless leader), it
         // should use requested, or default `compute_unit_limit` as transaction's execution cost.
         if !bank
             .feature_set
             .is_active(&feature_set::apply_cost_tracker_during_replay::id())
         {
-            QosService::update_or_remove_transaction_costs(
+            QosService::update_costs(
```

Comment:
+1 for only applying feature gate to committed transactions. 

Side note, transactions are not recorded or committed, include AccountInUse, InvalidFeePayer, TooManyAccountLocks, InsufficientFundsForFee etc are only dealt by leader at `Remove_cost()` above. Only those committed will be seen by replay_stage and subject to cost tracking. This makes sure logic gated by feature on leader and replay sides are consistent. 

Maybe good to add tests to check these changes, can do the tests separately tho.